packages/kotti-ui/source/props.ts

```diff
@@ -0,0 +1,198 @@
+ *   should mean that annotating the prop types explicitly is no longer necessary) in `@vue/composition-api`
+ *   (`0.6.1` and `1.1.5`), `PropType<>` appears to be bugged.
+ *   For example KtBreadcrumb’s props get inferred to be any
+ *   So, unfortunately, this may only help in the future, once the Vue team fixes this
+ * 3. `z.function().default()` does not accept `undefined` while parsing the schema manually
+ *   without an explicit `.optional()` {@link https://github.com/colinhacks/zod/issues/647}
+ *   However, it can still be used to generate the Vue defaults, as Vue doesn’t call validators with `undefined`
+ *
+ * @example
+ * // KtUserMenu.vue
+ * export default defineComponent<KottiUserMenu.PropsInternal>({
+ *   name: 'KtUserMenu',
+ *   props: makeProps(KottiUserMenu.propsSchema)
+ * })
+ *
+ * // types.ts
+ * export namespace KottiUserMenu {
+ *   const propsSchema = z.object({
+ *     // ...
+ *   })
+ *   type Props = z.input<typeof propsSchema>
+ *   type PropsInternal = z.output<typeof propsSchema>
+ * }
+ */
+export const makeProps = <SCHEMA extends z.ZodObject<z.ZodRawShape>>(
+	schema: SCHEMA,
+): {
+	[KEY in keyof SCHEMA['shape']]: Omit<PropOptions, 'required' | 'type'> & {
+		required: z.output<SCHEMA>[KEY] extends undefined ? false : true
+		type: PropType<z.output<SCHEMA>[KEY]>
+	}
+} =>
+	Object.fromEntries(
+		Object.entries(schema.shape).map(([key, shape]) => {
+			/* eslint-disable no-console */
+			if (DEBUG_MAKE_PROPS) console.log(`makeProp: generating “${key}”`)
+
+			const zodTypeSet = walkShape(shape)
+
+			/**
+			 * WORKAROUND: Usually, one should probably call shape.isOptional()
+			 *
+			 * However, this currently doesn’t seem to work with z.function().default()
+			 * and this also internally just calls safeParse on the object — evaluating
+			 * the default function, which wastes performance AND could cause
+			 * unintended side-effects
+			 */
+			const isOptional =
+				zodTypeSet.has('ZodDefault') || zodTypeSet.has('ZodOptional')
+
+			const prop: PropOptions<unknown, boolean> = {
+				validator: propValidator(shape, key),
+			}
+
+			if (isOptional) prop.default = shape._def.defaultValue
+			else prop.required = true
+
+			const propType = [...zodTypeSet]
+				.filter((x) => !ignoredZodTypes.has(x))
+				.map((zodTypeName) => {
+					if (DEBUG_MAKE_PROPS)
+						console.log(
+							`makeProps: found type ${zodTypeName} (${typeof zodTypeName})`,
+						)
+
+					if (!zodToVueTypes.has(zodTypeName))
+						throw new Error(`makeProps: unknown zod type “${zodTypeName}”`)
+
+					return zodToVueTypes.get(zodTypeName) as MapValueType<
+						typeof zodToVueTypes
+					>
+				})
+
+			prop.type = propType.length === 1 ? propType[0] : propType
+
+			return [key, prop]
+			/* eslint-enable no-console */
+		}),
+	) as {
+		[KEY in keyof SCHEMA['shape']]: Omit<PropOptions, 'required' | 'type'> & {
+			required: z.output<SCHEMA>[KEY] extends undefined ? false : true
+			type: PropType<z.output<SCHEMA>[KEY]>
```

Comment:
wouldn't `type` here be the same thing that was excluded from PropType
so
```Exclude PropType<z.output<SCHEMA>[KEY]>, Array<unknown>```

and as proposed in a separate comment
`PropConstructor<z.output<SCHEMA>[KEY]>`

Reply:
No, because of this line: https://github.com/3YOURMIND/kotti/pull/504/files#diff-fbcbf2aa3c29073ef5702e2fef52a3667008fcdc1a0b78c3825ec65bdbee1ed7R188